### PR TITLE
Define interfaces for common job operators

### PR DESCRIPTION
**Pull request by jian-he**
Creating the interfaces for custom controllers to implement, that is:

```
	// DeleteJob deletes the job
	DeleteJob(job interface{}) error

	// UpdateJobStatus updates the job status and job conditions
	UpdateJobStatus(job interface{}, replicas map[common.ReplicaType]*common.ReplicaSpec, jobStatus common.JobStatus) error

	// CreateService creates the service
	CreateService(job interface{}, service *v1.Service) error

	// DeleteService deletes the service
	DeleteService(job interface{}, service *v1.Service) error

	// CreatePod creates the pod
	CreatePod(job interface{}, podTemplate *v1.PodTemplate) error

	// DeletePod deletes the pod
	DeletePod(job interface{}, pod *v1.Pod) error

	// SetClusterSpec sets the cluster spec for the pod
	SetClusterSpec(job interface{}, podTemplate *v1.PodTemplate, rtype, index string) error
}
```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/common/12)

@richardsliu @johnugeorge @terrytangyuan

### CODE REVIEW

File: job_controller/job_controller.go [hunk 1]
```diff
@@ -60,6 +61,27 @@ type ControllerInterface interface {
 
 	// Returns the Job from API server
 	GetJobFromAPIClient(namespace, name string) (metav1.Object, error)
+
+	// DeleteJob deletes the job
+	DeleteJob(job interface{}) error
```
Comment (gaocegege) [inline job_controller/job_controller.go]: maybe runtime.Object is better?

Comment (johnugeorge) [reply job_controller/job_controller.go]: And CreateJob missing?

Comment (jian-he) [reply job_controller/job_controller.go]: Some methods require both "metav1.Object" (for things like GetName,GetNamespace) and "runtime.Object" (for ".Recorder.Eventf")types, so thought it's better to be make it consistent to be interface{}

Comment (gaocegege) [reply job_controller/job_controller.go]: SGTM

Comment (gaocegege) [reply job_controller/job_controller.go]: Yeah, same question with @johnugeorge 

Can we define CreateJob here?

Comment (jian-he) [reply job_controller/job_controller.go]: CreateJob is actually done by the user when creating a CRD (i.e. TFJob). I think operators will not create its own CRDs ?

Comment (gaocegege) [reply job_controller/job_controller.go]: Yeah, sure, I think so, thanks.

**Pull request merged**